lookout/style/typos/preprocessing/metrics.py

```diff
@@ -0,0 +1,194 @@
+from functools import partial
+from typing import NamedTuple
+
+import numpy
+import pandas
+
+from lookout.style.typos.utils import COLUMNS
+
+
+Scores = NamedTuple("Scores", (("tp", int),
+                               ("fp", int),
+                               ("tn", int),
+                               ("fn", int)))
+
+
+def detection_score(data: pandas.DataFrame, suggestions: dict) -> Scores:
+    """
+    Calculate score of solution for typo detection problem.
+
+    Typo is detected right: token is corrected then and only then when the token is not typoed.
+    Token is corrected: the first suggestion doesn't match the token.
+    :param data: DataFrame which indexed by "id" and has columns "typo", "corrupted".
+    :param suggestions: {id : [(candidate, correct_prob)]}, candidates are sorted
+                        by correct_prob in a descending order .
+    :return: Scores of suggestions on detection problem.
+    """
+    scores = Scores(0, 0, 0, 0)
+    for i in data.index:
+        if data.loc[i, COLUMNS["TOKEN"]] != data.loc[i, COLUMNS["CORRECT_TOKEN"]]:
+            if suggestions[i][0][0] != data.loc[i, COLUMNS["TOKEN"]]:
+                scores.tp += 1
+            else:
+                scores.fn += 1
+        else:
+            if suggestions[i][0][0] == data.loc[i, COLUMNS["TOKEN"]]:
+                scores.tn += 1
+            else:
+                scores.fp += 1
+    return scores
+
+
+def first_k_set(corrections: list, k: int) -> set:
+    """
+    Get set of k most probable correction tokens.
+
+    :param corrections: List of corrections, sorted by probability.
+    :param k: Number of corrections to take.
+    :return: Set of k most probable correction tokens.
+    """
+    first_k = set()
+    for correction, _prob in corrections[:k]:
+        first_k.add(correction)
+    return first_k
+
+
+def score_at_k(data: pandas.DataFrame, suggestions: dict, k: int = 1) -> Scores:
+    """
+    Calculate score@k of solution for typo correction problem.
+
+    The suggestions for typo correction are considered right
```

Comment:
```suggestion
    The suggestions for typo correction are considered correct
```
(more common word in this case)